linera-base/src/data_types.rs

```diff
@@ -880,20 +875,51 @@ pub struct CompressedBytecode {
     pub compressed_bytes: Vec<u8>,
 }
 
+#[cfg(not(target_arch = "wasm32"))]
 impl CompressedBytecode {
+    /// Returns `true` if the decompressed size does not exceed the limit.
+    pub fn decompressed_size_at_most(&self, limit: u64) -> Result<bool, DecompressionError> {
+        let mut decoder = zstd::stream::Decoder::new(&*self.compressed_bytes)?;
+        let limit = usize::try_from(limit).unwrap_or(usize::MAX);
+        let mut writer = LimitedWriter::new(io::sink(), limit);
+        match io::copy(&mut decoder, &mut writer) {
+            Ok(_) => Ok(true),
+            Err(error) => {
+                error.downcast::<LimitedWriterError>()?;
+                Ok(false)
+            }
+        }
+    }
+
     /// Decompresses a [`CompressedBytecode`] into a [`Bytecode`].
-    #[cfg(not(target_arch = "wasm32"))]
     pub fn decompress(&self) -> Result<Bytecode, DecompressionError> {
         #[cfg(with_metrics)]
         let _decompression_latency = BYTECODE_DECOMPRESSION_LATENCY.measure_latency();
-        let bytes = zstd::stream::decode_all(&*self.compressed_bytes)
-            .map_err(DecompressionError::InvalidCompressedBytecode)?;
+        let bytes = zstd::stream::decode_all(&*self.compressed_bytes)?;
 
         Ok(Bytecode { bytes })
     }
+}
+
+#[cfg(target_arch = "wasm32")]
+impl CompressedBytecode {
+    /// Returns `true` if the decompressed size does not exceed the limit.
+    pub fn decompressed_size_at_most(&self, limit: u64) -> Result<bool, DecompressionError> {
+        let compressed_bytes = &*self.compressed_bytes;
+        let limit = usize::try_from(limit).unwrap_or(usize::MAX);
+        let mut writer = LimitedWriter::new(io::sink(), limit);
+        let mut decoder = ruzstd::streaming_decoder::StreamingDecoder::new(compressed_bytes)
+            .map_err(io::Error::other)?;
+        match io::copy(&mut decoder, &mut writer) {
```

Comment:
Does it properly decompress even under the existence of the issue linked below? i.e. that it decompresses only a single frame.

Reply:
No, that issue is not fixed by this PR.

Reply:
Right, my question is - does `io::copy` decompress the `compressed_bytes` properly? i.e. does not end after first frame? b/c IIUC you need to handle that properly as well to detect the error here.

Reply:
Good point, I'll copy the `TODO` comment here, too.

I'd expect `io::copy` to fail in the same way, after the first frame, yes.

Reply:
Yes, that's what I meant. Alternatively, you could call `self.decompress` and check the length of the returned value.

Reply:
I wanted to use `io::sink()` to avoid allocating the memory for the decompressed data that we don't need.

Reply:
I was also considering extracting the decoder, i.e. `Write`r, creation, and only put that behind the feature flags, deduplicating everything else.

But I'm not sure what the solution to #2710 will look like in detail, and whether there will still be a single writer object. So if we simplify this I'd rather do that after the fix.